### ISSUE TITLE
Use prismarine-world

### ISSUE THREAD

**Issue by rom1504** [andrewrk/mineflayer#334]
This will make blockAt async and break everything depending on it.
But it's for the good cause of being able to implement a world representation with any kind of backend (especially async backend like databases)

https://github.com/PrismarineJS/prismarine-world

https://github.com/andrewrk/mineflayer/blob/master/lib/plugins/blocks.js

**Comment by rom1504** [comment]
One way to make the transition smoother is to just deprecate blockAt, implement it using prismarine-world-sync and use bot.world for everything else.

**Comment by rom1504** [comment]
Something interesting about this is a prismarine-world-blockfinder could be made and used in mineflayer-blockfinder that would then be just a few lines.

**Comment by rom1504** [comment]
requires https://github.com/PrismarineJS/prismarine-chunk/issues/16#issuecomment-169614381

**Comment by rom1504** [comment]
(using prismarine-world-sync would break mineflayer in the browser though, maybe it should be an optional dependency and we'd have no blockAt function for the no-prismarine-world-sync case, or maybe we could just remove blockAt altogether, but that would mean doing PRs on all projects using mineflayer)

**Comment by rom1504** [comment]
Related idea : have a prismarine-world-snapshot that would take a snapshot of a given region and allow a sync api on that region. That might be useful to speed up blockfinder.
(or maybe doing it asyncly would be fast enough, better benchmark before doing that)

**Comment by rom1504** [comment]
Alright, let's start with using prismarine-chunk in mineflayer. That will give us cross platform chunk (https://github.com/PrismarineJS/prismarine-chunk/issues/19) + we don't have to deal with the async problem for now.

**Comment by rom1504** [comment]
p-chunk is now used.
using prismarine-world could be considered but would involve a lot of async everywhere. Not not obvious to do

**Comment by rom1504** [comment]
https://github.com/PrismarineJS/mineflayer/pull/1145 made this happen

Follow up : 
* [ ] move find block to somewhere outside of mineflayer, so it can be used with any prismarine-world (either its own package, either directly in prismarine-world)
* [ ] make the viewer use prismarine-world only
* [ ] create an example showing how to save a mineflayer world to anvil (rendered trivial by using prismarine-world)
* [ ] A way to store block entities https://github.com/PrismarineJS/mineflayer/blob/master/lib/plugins/blocks.js#L27 it would make sense to have it directly in prismarine-chunks similar to what mojang do (would simplify things for storing chunks), but that could go first in prismarine-world
* [ ] store signs paintingsByPos and paintingsById in prismarine-world (/chunk)
* [ ] Some events, world is already an EventEmitter, but no events are emitted (blockUpdate, chunkColumnLoad/Unload, etc..) eventually those events could be removed or proxyed in mineflayer
* [ ] entities ? If we want to be able to use only pworld in pviewer, we also need the world to emit entity related events: https://github.com/PrismarineJS/prismarine-viewer/blob/master/lib/mineflayer.js#L82
* [ ] deprecate bot.blockAt , bot.world.getBlock should be used. (that'll be when everything of the world is ported to prismarine-world)

**Comment by rom1504** [comment]
deprecating blockAt would be really useful

**Comment by introquest** [comment]
Faced with the fact that **mineflayer** duplicates some things from **prusmarine** such as _Block_, _Item_, _Recipe_, etc.
Is it about reusing all of this from **prismarine**?
And what remains to be done on this task?

**Comment by TheDudeFromCI** [comment]
@introquest Mineflayer does not duplicate anything from those libraries, but imports then. The goal of Mineflayer is to move as much code out of Mineflayer as possible into micro-modules. (prismarine-block, primsarine-item, etc) Ideally, by making many very small libraries, they become very easy to maintain and reuse since they are so small.

As for this issue, prismarine-world should contain all of the world states, with Mineflayer acting solely as a stateless transformer for the world. The world contains submodules that further divide the state. Prismarine-chunk, Prismarine-Entity, etc.

**Comment by rom1504** [comment]
signs and block entities are done

next step deciding what to do with paintings

**Comment by rom1504** [comment]
once paintings is gone we can merge https://github.com/PrismarineJS/mineflayer/pull/2453

**Comment by rom1504** [comment]
new plan after discussion in discord:

move everything to pchunk
* [ ] paintings
* [ ] entities
* [ ] inventories as part of entities
* [ ] players as part of entities
* [ ] bots as part of players

since entities can move between chunks, 
* [ ] pworld should handle the move of entities from one chunk to the next

once this is done, mineflayer should contain no state and be only a packet <-> world and action <-> packet/world synchronization layer
it will allow
* same state storage for bot/client and server
* natural multi bot support with one world
* possibility to emulate world changes without client and server
* easy to test features of the world (like find block, pathfinding, visualization...) without network

**Comment by rom1504** [comment]
since we have some funding now, any significant PR towards that goal is eligible to $20 of reward (using open collective money)

of course working on it because it's cool stays a great reason :)

**Comment by Gjum** [comment]
why would entities be part of a chunk, and not just part of the world? their id is unique in the entire world, most entities move freely between chunks all the time, and accessing entities through their current chunk is rather useless as you typically want to find entities nearby regardless of what chunk they are in.

**Comment by rom1504** [comment]
the world abstraction (pworld) would expose the ability to search for an entity by its id or to look for an entity in a location based on properties (by looking and/or indexing the information stored in chunks)
having the entities in the chunk make it natural to save/load them to disk (anvil) and to load/unload them when chunks get loaded/unloaded. It also does a clean separation of concern between chunk storing the data, and world just exposing and doing the orchestration

having the entities be part of the chunk should make it easier to scale the world to large sizes

**Comment by rom1504** [comment]
From Claude:


Looking at the current state of the codebase, I have a proposal for making incremental progress on issue #334 without breaking existing functionality.

## Current Analysis

**prismarine-world**: Already has sync APIs (`WorldSync` class) and stores world state
**prismarine-chunk**: Already stores `blockEntities` (chest inventories, furnace contents, etc.)  
**mineflayer**: Currently manages block inventories separately from the world

## Minimal Action Plan

**Phase 1: Consolidate Block Inventories**

Move block-attached inventories (chests, furnaces, hoppers, etc.) from mineflayer's separate tracking into prismarine-world's existing sync interface.

**Current:**
```javascript
const chest = await bot.openChest(chestBlock) // mineflayer manages inventory
```

**Proposed:**
```javascript
const inventory = bot.world.getBlockEntity(pos).inventory // inventory lives in world
```

**Benefits:**
- Uses existing `prismarine-chunk.blockEntities` storage
- No breaking changes to `blockAt()` API
- Consolidates world state in one place  
- Small step toward full world state unification
- Makes packaging/modularity better

**Next phases** could then move entities, then eventually remove async APIs if desired, but this gets us moving without "breaking everything."

This leverages the infrastructure that's already there and makes concrete progress on consolidating world state into prismarine-world.

**Comment by rom1504** [comment]
**Update to my previous comment** - I need to correct my understanding:

## Corrected Analysis

Looking at flying-squid, I see that:
- **Player inventories** belong to individual players (each player has )
- **Block inventories** (chests, furnaces, etc.) are world state that multiple players can access
- prismarine-chunk already stores  but doesn't handle the inventory UI/interaction layer

## Revised Minimal Plan

**Phase 1: Move Block Inventories to World**

Move block-attached inventories (chests, furnaces, hoppers, etc.) from mineflayer's separate tracking into prismarine-world, while keeping player inventories with players.

**Current separation:**


**Proposed:**


This way:
- Player inventories remain per-player (correct for multi-player servers)
- Block inventories move to world state (can be shared between players)
- Uses existing prismarine-chunk blockEntities infrastructure

This makes more sense for the flying-squid use case where N players share one world but have separate inventories.

**Comment by rom1504** [comment]
**Updated Plan: Store ALL Inventories in prismarine-world**

Looking at the code, prismarine-world currently only stores chunks/blocks but not entities. Here's how to store ALL inventories in prismarine-world:

## Storage Strategy

**Add entity storage to prismarine-world:**

```javascript
class World {
  constructor() {
    this.columns = {}        // existing chunk storage
    this.entities = new Map() // NEW: entity storage by UUID
    this.blockEntities = {}  // move from chunks to world level
  }
  
  // NEW entity methods
  setEntity(uuid, entityData) { /* store player/entity with inventory */ }
  getEntity(uuid) { /* retrieve player/entity */ }
  removeEntity(uuid) { /* cleanup */ }
  
  // Enhanced block entity methods  
  setBlockEntity(pos, inventory) { /* chest/furnace inventories */ }
  getBlockEntity(pos) { /* retrieve block inventory */ }
}
```

**Migration path:**
1. **Block inventories**: Move from mineflayer's separate tracking to `world.getBlockEntity(pos).inventory`
2. **Player inventories**: Move from `player.inventory` to `world.getEntity(player.uuid).inventory` 
3. **Entity inventories**: Horses, llamas, etc. all stored as `world.getEntity(entityId).inventory`

**Benefits:**
- Single source of truth for ALL world state in prismarine-world
- Works for both single-player (mineflayer) and multi-player (flying-squid)
- Enables better packaging/modularity as requested in the issue
- Clean separation: mineflayer becomes thin client, prismarine-world handles state

**Implementation order:**
1. Add entity storage infrastructure to prismarine-world
2. Move block inventories first (smaller change)  
3. Move player inventories (bigger breaking change)

This puts the entire world state stack (blocks, entities, inventories) in prismarine-world as requested.

**Comment by extremeheat** [comment]
wrt block entities, prismarine-chunk tries to handle that internally (it reads block entities when a getBlock is called), so some changes may need to be made there to preload all the block entities inside pworld for the relevant chunk

**Comment by DragonModels** [comment]
I'm evaluating a contribution with disclosed AI assistance. I reviewed #3737 and PrismarineJS/prismarine-world#157; those focus on entity storage, while the block-inventory step discussed here appears separate.

Would a focused contribution exposing observed chest inventory state through prismarine-world, while preserving Mineflayer's existing window API, fit the current direction? I would distinguish contents actually received from the server from unknown/stale contents and test slot updates and chunk unloads locally; I have not implemented or validated this yet.

Is the previously offered $20 reward for significant progress still available for an accepted contribution of that scope, with payout through Open Collective? No assignment or bounty claim is being made.
